### PR TITLE
Parse error recovery and incrementality for GHC

### PR DESCRIPTION
**Pull request by sgraf812**

### CODE REVIEW

File: content/ideas/parse-error-recovery.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+been [tried on GHC as a proof of concept](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/11990).
+
+The goal of this project is to take over the pull request to `happy` so that it
+can be merged, and then use the improved `happy` to generate multiple and better
+parse error messages in GHC.
+
+There are a couple of stretch goals:
+
+  * `happy` could further be improved to pass a closure of its
+    parse state to reduction actions, so as to enable incremental parsing.
```
Comment (michaelpj) [inline content/ideas/parse-error-recovery.md:29]: ... and use this in GHC?

File: content/ideas/parse-error-recovery.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  * Improve `happy` so that it provides a convenient and encapsulated way to
+    introspect the LALR item stack, for example to identify bracketing
```
Comment (michaelpj) [inline content/ideas/parse-error-recovery.md:31]: ... and use this in GHC?

File: content/ideas/parse-error-recovery.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    productions such as `'(' expr . ')'` in order to report mismatched brackets.
+    There is a [hacky GHC Merge Request](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4711) 
+    that tries to achieve as much without buy in from happy.
+  * Improve `happy`s code base, which by now is over 25 years old.
```
Comment (michaelpj) [inline content/ideas/parse-error-recovery.md:35]: Might be helpful to identify a couple of concrete things: just "modernisation", or comments, or tests, or what?

**Review by michaelpj — APPROVED** [review]

**Comment by aaronallen8455** [comment]
LGTM, thanks! I'm going to go ahead and merge this so that it's listed on the site before I submit the GSoC application. Feel free to open another PR if you want to make any changes.

File: content/ideas/parse-error-recovery.md (hunk 3 above)
Comment (sgraf812) [reply content/ideas/parse-error-recovery.md:35]: Indeed it is rather hard to spell out what could make a substantial improvment. Certainly it appears that the modularisation effort has left `happy` in a rather rough stage. See #183.

**Comment by sgraf812** [comment]
Thanks! I addressed Michael's feedback in #183.

**Pull request merged**